luigi/s3.py

```diff
@@ -584,6 +586,180 @@ def move_to_final_destination(self):
         self.s3_client.put_multipart(self.tmp_path, self.path, **self.s3_options)
 
 
+class AtomicRemoteWritableS3File(object):
+    """
+    An S3 file that writes to a remote temp object on S3; copies to the true key on close.
+
+    This class requires boto3 v1.4.0+ for its non-seekable file object upload ability.
+
+    Useful for performing operations on large S3 objects when you don't have
+    sufficient space on local drives.
+
+    Works around AWS S3's multipart transfer size requirements and boto3's
+    idiosyncratic implementation that requires an initial buffer size
+    larger than the multipart transfer threshold in order to correctly
+    select the 'read-until-empty' behavior needed for a streaming upload.
+    """
+    _boto3_default_multipart_threshold = 8 * 1024 * 1024
+
+    def __init__(self, s3_bucket, s3_key, boto3_s3_client=None):
+        import boto3
+        from s3transfer.manager import TransferManager, TransferConfig
```

Comment:
yeah, this is the reason. It's certainly a tradeoff, but I've seen the pattern elsewhere in the luigi codebase. See RemoteFileSystem in https://github.com/spotify/luigi/blob/master/luigi/contrib/ftp.py for an example where pysftp is not a hard requirement, because some users of the RemoteTarget will only care about FTP and won't have pysftp installed.